network.go

```diff
@@ -0,0 +1,505 @@
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package firecracker
+
+import (
+	"context"
+	"net"
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"github.com/containernetworking/cni/libcni"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+
+	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
+	"github.com/firecracker-microvm/firecracker-go-sdk/cni/vmconf"
+)
+
+const (
+	defaultCNIBinDir   = "/opt/cni/bin"
+	defaultCNIConfDir  = "/etc/cni/conf.d"
+	defaultCNICacheDir = "/var/lib/cni"
+)
+
+// NetworkInterface represents a Firecracker microVM's network interface.
+// It can be configured either with static parameters set via StaticConfiguration
+// or via CNI as set via CNIConfiguration. It is currently an error to specify
+// both static and CNI configuration.
+type NetworkInterface struct {
+	// StaticConfiguration parameters that will be used to configure the VM's
+	// tap device and internal network for this network interface.
+	StaticConfiguration *StaticNetworkConfiguration
+
+	// CNIConfiguration that will be used to generate the VM's network namespace,
+	// tap device and internal network for this network interface.
+	CNIConfiguration *CNIConfiguration
+
+	// AllowMMDS makes the Firecracker MMDS available on this network interface.
+	AllowMMDS bool
+
+	// InRateLimiter limits the incoming bytes.
+	InRateLimiter *models.RateLimiter
+
+	// OutRateLimiter limits the outgoing bytes.
+	OutRateLimiter *models.RateLimiter
+}
+
+// CNIConfiguration specifies the CNI parameters that will be used to generate
+// the network namespace and tap device used by a Firecracker interface.
+//
+// Currently, CNIConfiguration can only be specified for VMs that have a
+// single network interface.
+type CNIConfiguration struct {
+	// NetworkName (required) corresponds to the "name" parameter in the
+	// CNI spec's Network Configuration List structure. It selects the name
+	// of the network whose configuration will be used when invoking CNI.
+	NetworkName string
+
+	// IfName (optional) corresponds to the CNI_IFNAME parameter as specified
+	// in the CNI spec. It generally specifies the name of the interface to be
+	// created by a CNI plugin being invoked.
+	//
+	// Note that this does NOT necessarily correspond to the name of the
+	// tap device the Firecracker VM will use as the tap device may be
+	// created by a chained plugin that adapts the tap to a pre-existing
+	// network device (which will by the one with "IfName").
+	IfName string
+
+	// Args (optional) corresponds to the CNI_ARGS parameter as specified in
+	// the CNI spec. It allows custom args to be passed to CNI plugins during
+	// invocation.
+	Args [][2]string
```

Comment:
Did we decide that changing this to `map[string]string` wasn't feasible?

Reply:
The problem is that the CNI spec doesn't say to my knowledge whether the order of the args matters or not. If we use a map we are assuming order doesn't matter, which may or may not be true for arbitrary CNI plugins.